File: services/admin-service/src/Models/AuditLog.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace OtterWorks.AdminService.Models;

[Table("audit_logs")]
public class AuditLog
{
    [Key]
    [Column("id")]
    public Guid Id { get; set; }

    [Column("actor_id")]
    public Guid? ActorId { get; set; }

    [Column("actor_email")]
    public string? ActorEmail { get; set; }

    [Required]
    [Column("action")]
    public string Action { get; set; } = string.Empty;

    [Required]
    [Column("resource_type")]
    public string ResourceType { get; set; } = string.Empty;

    [Column("resource_id")]
    public Guid? ResourceId { get; set; }

    [Column("changes_made", TypeName = "jsonb")]
    public string ChangesMade { get; set; } = "{}";

    [Column("ip_address")]
    public string? IpAddress { get; set; }

    [Column("user_agent")]
    public string? UserAgent { get; set; }

    [Column("created_at")]
    public DateTime CreatedAt { get; set; }

    [Column("updated_at")]
    public DateTime UpdatedAt { get; set; }
}

File: services/admin-service/app/models/audit_log.rb
class AuditLog < ApplicationRecord
  ACTIONS = %w[
    user.created user.updated user.suspended user.activated user.deleted
    feature_flag.created feature_flag.updated feature_flag.deleted
    config.updated announcement.created announcement.updated announcement.deleted
    quota.updated bulk.users_updated
  ].freeze

  validates :action, presence: true
  validates :resource_type, presence: true

  scope :by_action, ->(action) { where(action: action) }
  scope :by_resource, lambda { |type, id = nil|
    scope = where(resource_type: type)
    scope = scope.where(resource_id: id) if id.present?
    scope
  }
  scope :by_actor, ->(actor_id) { where(actor_id: actor_id) }
  scope :recent, -> { order(created_at: :desc) }
  scope :since, ->(time) { where('created_at >= ?', time) }

  RECORDABLE_DETAILS = %i[resource_id actor_id actor_email changes_made ip_address user_agent].freeze

  def self.record!(action:, resource_type:, **details)
    unknown = details.keys - RECORDABLE_DETAILS
    raise ArgumentError, "unknown keyword#{'s' if unknown.many?}: #{unknown.join(', ')}" if unknown.any?

    create!(
      details.reverse_merge(changes_made: {})
             .merge(action: action, resource_type: resource_type)
    )
  end
end
